bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BVM.java

```diff
@@ -4403,12 +4403,8 @@ private static void insertToMap(Strand ctx, BMap bMap, String fieldName, BValue
     }
 
     private static boolean isValidMapInsertion(BType mapType, BValue value) {
-        if (value == null) {
-            return true;
-        }
-
         BType constraintType = ((BMapType) mapType).getConstrainedType();
-        if (constraintType == BTypes.typeAny || constraintType.equals(value.getType())) {
+        if (value != null && constraintType.equals(value.getType())) {
```

Comment:
The first check in `checkIsType` (after checking whether the value is mutable) is `sourceType == targetType || sourceType.equals(targetType)`. So I guess we can remove this check and just call `checkIsType`, WDYT? We can do it separately maybe.